### PR TITLE
Add report command

### PR DESCRIPTION
**Pull request by nitanmarcel**
Contains requested improvements from https://github.com/Lonami/TelethonianBotExt/pull/33.

- A cooldown between reports per chat (Defaults to 10 minutes)
- A limit on how many times a message can be reported (Defaults to 1)
- Only allow reporting messages with media, usernames, or links.

### CODE REVIEW

**Comment by nitanmarcel** [comment]
One thing I'll have to fix:

- The script checks the report time on the first report, if the cooldown passed and another message is reported the last report won't be taken in consideration. The fix should be fixed easily only by getting the last element in the list.


Tell me if there are more things to fix

**Comment by nitanmarcel** [comment]
Ok, the last commit fixed everything and simplified a little how things work. Also thanks to @painor for helping me with suggestions

@Lonami the PR is ready to be merged if you don't think anything else needs to be changed ^^

File: reports.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+async def init(bot: TelegramClient):
+    COOLDOWN = 10 * 60
+    REPORTS: Dict[int, ReportedMessages] = {}
+    MAX_N_REPORTS = 5
+
+    @bot.on(
+        NewMessage(
+            pattern=r"^(#|\/)report", func=lambda e: not e.is_private and e.is_reply
+        )
+    )
```
Comment (Lonami) [inline reports.py:33]:
```suggestion
    @bot.on(NewMessage(
        pattern=r"^(#|\/)report",
        func=lambda e: not e.is_private and e.is_reply
    ))
```

File: reports.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                        MessageEntityMention,
+                        MessageEntityMentionName,
+                        MessageEntityTextUrl,
+                        MessageEntityUrl,
+                    ),
+                )
+                for entity in (reply_message.entities or ())
+            )
+            and not bool(event.file)
+        ):
```
Comment (Lonami) [inline reports.py:53]:
```suggestion
        if not (event.file or any(isinstance(entity, (
            MessageEntityEmail,
            MessageEntityMention,
            MessageEntityMentionName,
            MessageEntityTextUrl,
            MessageEntityUrl,
        )) for entity in (reply_message.entities or ()))):
```

File: reports.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            await event.delete()
+            return
+
+        if reports:
+            if (time.time() - reports.last_time) < COOLDOWN:
+                await event.delete()
+                return
+            if reply_message.id in reports.reported_messages[:MAX_N_REPORTS]:
```
Comment (Lonami) [inline reports.py:61]: `[:MAX_N_REPORTS]` won't be necessary once we use a bounded deque.

File: reports.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            event.chat_id, filter=ChannelParticipantsAdmins
+        ):
+            admin: User
+            if not admin.bot:
+                try:
+                    await bot.send_message(
+                        admin.id,
+                        f"[{sender.first_name}](t.me/{sender.id}) reported a [message](t.me/{chat.username}/{reply_message.id}) in [{chat.title}](t.me/{chat.username})",
+                    )
+                except UserIsBlockedError:
```
Comment (Lonami) [inline reports.py:84]: I think the error is different if a user never talked to the bot first?

**Review by Lonami — CHANGES_REQUESTED** [review]
Will review logic after the changes I requested are made.

**Comment by nitanmarcel** [comment]
@Lonami tell me if there's something else :)

**Comment by nitanmarcel** [comment]
Actually if you could make a small change commit to `is_index_at_max` to check if the id is in the deque instead of getting the index it will be nice since I will be able to do a fast commit from my phone ^^

The `popleft` function came in my mind before I closed my laptop and I forgot to change the respective function :)

Thanks

File: reports.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+from collections import deque
+
+from telethon import TelegramClient
+from telethon.errors import UserIsBlockedError
+from telethon.events import NewMessage
+from telethon.tl.custom import Message
+from telethon.tl.types import (Channel, ChannelParticipantsAdmins, Chat,
+                               MessageEntityEmail, MessageEntityMention,
+                               MessageEntityMentionName, MessageEntityTextUrl,
+                               MessageEntityUrl, User)
```
Comment (Lonami) [inline reports.py:12]:
```suggestion
from telethon.tl.types import (
    Channel, ChannelParticipantsAdmins, Chat, MessageEntityEmail,
    MessageEntityMention, MessageEntityMentionName, MessageEntityTextUrl,
    MessageEntityUrl, User
)
```

File: reports.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    def __init__(self, max_len: int):
+        self.reported_messages: Deque[int] = deque()
+        self.max_len = max_len
+        self.last_time: float = 0.0
+
+    def add(self, msg_id: int):
+        self.reported_messages.append(msg_id)
+        self.last_time = time.time()
+        if len(self.reported_messages) > self.max_len:
+            self.reported_messages.popleft()
```
Comment (Lonami) [inline reports.py]: Not necessary after we use deque as intended.

```suggestion
```

File: reports.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+                event.chat_id,
+                filter=ChannelParticipantsAdmins):
+            admin: User
+            if not admin.bot:
+                try:
+                    await bot.send_message(
+                        admin.id,
+                        f"[{sender.first_name}](t.me/{sender.id}) reported a [message](t.me/{chat.username}/{reply_message.id}) in [{chat.title}](t.me/{chat.username})",
+                    )
+                except (UserIsBlockedError, ValueError):
```
Comment (Lonami) [inline reports.py:80]: Are you sure the error here is a `ValueError`?

Comment (nitanmarcel) [reply reports.py:80]: Yes, I've tested it using your own user id which was never logged by the bot

**Comment by nitanmarcel** [comment]
I modified the deque to have a maxlen and commited it a few moments after your request

**Pull request merged**